Enable Compiler Warnings in a CI job

mabruzzo opened the pull request:
This is mostly being done in preparation for transcription in the newchem-cpp branch.

Background
----------
As we all know, it is very easy to write C/C++ code that is accepted by the compiler, but is wrong in a subtle manner. For example, C/C++ do lots of implicit conversions (which can lead to unexpected results).

Goal
----
The goal of this PR is to enable warning flags for the compiler and treat them as errors. We want to do this in a manner that isn't a burden to our users. In other words, we want these warnings (and their treatment as errors) to be off by default and to only turn them on in CI.

How We Accomplish This
----------------------
To accomplish this, I made a CMakePresets.json file, where we can store various presets for configuring Grackle-builds. It's designed to give you lots of customizability for reusing configuration-snippets. This will be very useful in the future if/when we try configuring sanitizers (to detect undefined behavior) and possibly for measuring code-coverage.

As an added bonus, it's also a great starting point for people who configuring IDEs (like vscode or CLion). People can make their own personal versions of the presets by creating a CMakeUserPresets.json file.

**Bottom Line:** this is meant to be a starting point. It would be great to merge this in ASAP so we can start using this for the C++ transcription PRs.

Documentation
-------------
Let me know if you want me to add narrative docs (I would only add these to the developer guide. Technically Presets require a newer version of CMake than our users use. But, I think this is okay for a subset of workflows run by CI)

brittonsmith commented:
I think this is a really good idea and happy to merge it. It looks like there are now some merger conflicts. If you can resolve those, I'm happy to see this go in. Have you tried compiling with these flags yet? I just assume that Grackle compiles with a non-zero number of warnings.

mabruzzo commented:
> I think this is a really good idea and happy to merge it. It looks like there are now some merger conflicts. If you can resolve those, I'm happy to see this go in. Have you tried compiling with these flags yet? I just assume that Grackle compiles with a non-zero number of warnings.

I have. I was surprised that Grackle "does so well." 
- In the initial version of the PR, I was able to silence one warning by making a tiny tweak in `initialize_chemistry_data.c`. And we have a warning in `calculate_dust_temperature.c` and `calculate_temperature.c` about unused variables (I'm holding off on fixing that to avoid merge conflicts).
- I actually added a few extra warnings a few days back (that I forgot to push before you reviewed the PR) that identified a few unused variables from `rate_functions.c` (I fixed that).
- After merging in with main, the warnings caught a few more unused variables within `cxx_omp_example.C` (they are now fixed).

In the future, I would like to also warn about implicit conversions, but that's going to be a task for after we catch up to the transcription PRs.

Let me know if you want to review the new changes or if I should go ahead and merge
